pkg/condition/condition.go

```diff
@@ -0,0 +1,109 @@
+package condition
+
+import (
+	"context"
+	"encoding/json"
+	"os"
+	"time"
+
+	operators "github.com/operator-framework/api/pkg/operators/v2"
+	"github.com/pkg/errors"
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kubeTypes "k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	patcher "github.com/openshift/windows-machine-config-operator/pkg/patch"
+)
+
+const (
+	upgradeableTrueReason   = "upgradeIsSafe"
+	upgradeableTrueMessage  = "The operator is safe for upgrade."
+	upgradeableFalseReason  = "upgradeIsNotSafe"
+	upgradeableFalseMessage = "The operator is currently processing sub-components."
+	// operatorConditionName is an environment variable set by OLM identifying the operator's OperatorCondition CR
+	OperatorConditionName = "OPERATOR_CONDITION_NAME"
+)
+
+// EnableOLMUpgrade is a wrapper that sets the Upgradeable condition to True to allow OLM to perform operator upgrades
+func EnableOLMUpgrade(c client.Client, watchNamespace string) error {
+	return set(c, watchNamespace, operators.Upgradeable, meta.ConditionTrue, upgradeableTrueReason,
+		upgradeableTrueMessage)
+}
+
+// DisableOLMUpgrade is a wrapper that sets the Upgradeable condition to False to prevent operator upgrades by OLM
+func DisableOLMUpgrade(c client.Client, watchNamespace string) error {
+	return set(c, watchNamespace, operators.Upgradeable, meta.ConditionFalse, upgradeableFalseReason,
+		upgradeableFalseMessage)
+}
+
+// set sets the given condition with the given values within the associated OperatorCondition.
+// No-op if operator is not managed by OLM
+func set(c client.Client, ocNamespace, conditionType string, status meta.ConditionStatus,
+	reason, message string) error {
+	// Check if operator is OLM-managed
+	if ocName, present := os.LookupEnv(OperatorConditionName); present {
+		oc, err := get(c, ocNamespace, ocName)
+		if err != nil {
+			return err
+		}
+		if err = patch(c, oc, conditionType, status, reason, message); err != nil {
+			return err
+		}
+		// Treat condition setting as a lock mechanism -- ensure spec change is actually picked up
```

Comment:
Not sure if I would call this a lock given a lock should work even if we are multi-threaded.

Reply:
I'll remove that part as I am not sure if it would work as intended with multithreading.